alica_engine/src/engine/Assignment.cpp

```diff
@@ -1,532 +1,448 @@
-        }
-    }
+    return AssignmentView(this, idx);
+}
 
-    auto robotsSucc = this->epSucMapping->getRobotsById(epid);
-    if (robotsSucc != nullptr) {
-        for (auto& robot : *robotsSucc) {
-            ret->push_back(robot);
-        }
-    }
+AssignmentView Assignment::getAgentsWorking(int64_t epId) const
+{
+    const EntryPoint* ep = _plan->getEntryPointByID(epId);
+    return ep ? AssignmentView(this, ep->getIndex()) : AssignmentView();
+}
 
-    return ret;
+AssignmentSuccessView Assignment::getAgentsWorkingAndFinished(const EntryPoint* ep) const
+{
+    return AssignmentSuccessView(this, ep->getIndex());
 }
 
-std::shared_ptr<SuccessCollection> Assignment::getEpSuccessMapping() const
+AllAgentsView Assignment::getAllAgents() const
 {
-    return this->epSucMapping;
+    return AllAgentsView(this);
+}
+AgentsInStateView Assignment::getAgentsInState(const State* s) const
+{
+    return AgentsInStateView(this, s);
 }
 
-void Assignment::setAllToInitialState(const AgentGrp& robots, const EntryPoint* defep)
+AgentsInStateView Assignment::getAgentsInState(int64_t sid) const
 {
-    auto rlist = this->epRobotsMapping->editRobotsByEp(defep);
-    for (const supplementary::AgentID* r : robots) {
-        rlist->push_back(r);
-        robotStateMapping->setState(r, defep->getState());
-    }
+    const State* s = _plan->getStateByID(sid);
+    return s ? AgentsInStateView(this, s) : AgentsInStateView();
 }
 
-bool Assignment::removeRobot(const supplementary::AgentID* robotId)
+void Assignment::clear()
 {
-    this->robotStateMapping->removeRobot(robotId);
-    AgentGrp* curRobots;
-    for (int i = 0; i < this->epRobotsMapping->getSize(); ++i) {
-        curRobots = this->epRobotsMapping->editRobots(i);
-        auto iter = find_if(curRobots->begin(), curRobots->end(), [&robotId](const supplementary::AgentID* id) { return *robotId == *id; });
-        if (iter != curRobots->end()) {
-            curRobots->erase(iter);
-            return true;
-        }
+    _successData.clear();
+    for (AgentStatePairs& asp : _assignmentData) {
+        asp.clear();
     }
-    return false;
 }
 
-void Assignment::addRobot(const supplementary::AgentID* id, const EntryPoint* e, const State* s)
+bool Assignment::updateAgent(AgentIDConstPtr agent, const EntryPoint* e)
 {
-    if (e == nullptr) {
-        return;
-    }
-    this->robotStateMapping->setState(id, s);
-    this->epRobotsMapping->addRobot(id, e);
-    return;
+    return updateAgent(agent, e, nullptr);
 }
 
-/**
- * Tests whether this assignment is valid with respect to the plan's cardinalities.
- * @return A bool
- */
-bool Assignment::isValid() const
+bool Assignment::updateAgent(AgentIDConstPtr agent, const EntryPoint* e, const State* s)
 {
-    const vector<shared_ptr<std::list<const supplementary::AgentID*>>>& success = this->epSucMapping->getRobots();
+    bool found = false;
+    bool inserted = false;
+    int i = 0;
+    assert(s == nullptr || s->getEntryPoint() == e);
```

Comment:
If we assume that e must be valid, why can't we also assume that s must be?

Reply:
Here, s == nullptr encodes a meaningful call, see a few lines above.